src/ucp/tag/rndv.c

```diff
@@ -5,19 +5,316 @@
 {
-    /* TODO */
-    ucs_fatal("unimplemented");
-    return UCS_ERR_NOT_IMPLEMENTED;
+    ucs_status_t status;
+
+    /* zcopy */
+    status = ucp_request_send_buffer_reg(req, ucp_ep_get_rndv_data_lane(req->send.ep));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ucp_ep_connect_remote(req->send.ep);
+
+    req->send.uct.func = ucp_proto_progress_rndv_rts;
+    return UCS_OK;
+}
+
+ucs_status_t ucp_proto_progress_rndv_get(uct_pending_req_t *self)
+{
+    ucp_request_t *get_req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucs_status_t status;
+    size_t offset, length;
+
+    if (ucp_ep_is_stub(get_req->send.ep)) {
+        return UCS_ERR_NO_RESOURCE;
+    }
+
+    ucs_debug("send rndv get");
+
+    /* reset the lane to rndv since it might have been set to 0 since it was stub on RTS receive */
+    get_req->send.lane = ucp_ep_get_rndv_data_lane(get_req->send.ep);
+
+    if (get_req->send.state.dt.contig.memh == UCT_INVALID_MEM_HANDLE) {
+        /* TODO Not all UCTs need registration on the recv side */
+        status = ucp_request_send_buffer_reg(get_req, get_req->send.lane);
+        ucs_assert_always(status == UCS_OK);
+
+        {
+            size_t max_get_zcopy = ucp_ep_config(get_req->send.ep)->max_rndv_get_zcopy;
+            size_t remainder = (uintptr_t) get_req->send.buffer % UCP_ALIGN;
+
+            if (remainder) {
+                get_req->send.uct_comp.count = 1 +
+                                               (get_req->send.length - (UCP_MTU_SIZE - remainder) +
+                                               max_get_zcopy - 1) / max_get_zcopy;
+            } else {
+                get_req->send.uct_comp.count = (get_req->send.length + max_get_zcopy - 1) / max_get_zcopy;
+            }
+        }
+    }
+
+    offset = get_req->send.state.offset;
+
+    ucs_debug("offset %zu remainder %zu", offset, (uintptr_t)get_req->send.buffer % UCP_ALIGN);
+
+    if ((offset == 0) && ((uintptr_t)get_req->send.buffer % UCP_ALIGN)) {
+        length = UCP_MTU_SIZE - ((uintptr_t)get_req->send.buffer % UCP_ALIGN);
+    } else {
+        length = ucs_min(get_req->send.length - offset,
+                         ucp_ep_config(get_req->send.ep)->max_rndv_get_zcopy);
+    }
+
+    ucs_debug("read %p len %zu", (void*)get_req->send.buffer + offset, length);
+    status = uct_ep_get_zcopy(ucp_ep_get_rndv_data_uct_ep(get_req->send.ep),
+                              (void*)get_req->send.buffer + offset,
+                              length,
+                              get_req->send.state.dt.contig.memh,
+                              get_req->send.rndv_get.remote_address + offset,
+                              get_req->send.rndv_get.rkey,
+                              &get_req->send.uct_comp);
+
+    if ((status == UCS_OK) || (status == UCS_INPROGRESS)) {
+        get_req->send.state.offset += length;
+        if (get_req->send.state.offset == get_req->send.length) {
+            return UCS_OK;
```

Comment:
yes, basically everything ucp_rndv_get_completion() does should be done here as well (probably use common function)